File: c#/money-problem.Tests/BankShould.cs
using FluentAssertions;
using money_problem.Domain;
using Xunit;
using static money_problem.Domain.Currency;

namespace money_problem.Tests
{
    public class BankShould
    {
        [Fact(DisplayName = "10 EUR -> USD = 12 USD")]
        public void ConvertEuroToUsd()
        {
            var bank = Bank.WithExchangeRate(EUR, USD, 1.2);

            bank.Convert(10d.Euros(), USD)
                .Should()
                .Be(12d.Dollars());
        }

        [Fact(DisplayName = "10 EUR -> EUR = 10 EUR")]
        public void ConvertMoneyInTheSameCurrency()
        {
            var bank = Bank.WithExchangeRate(EUR, USD, 1.2);

            bank.Convert(10d.Euros(), EUR)
                .Should()
                .Be(10d.Euros());
        }

        [Fact(DisplayName = "Throw missing exchange rate exception in case of missing exchange rates")]
        public void ConvertWithMissingExchangeRateShouldThrowException()
        {
            var portfolio = Bank.WithExchangeRate(EUR, USD, 1.2);

            portfolio.Invoking(p => p.Convert(10d.Euros(), KRW))
                .Should()
                .Throw<MissingExchangeRateException>()
                .WithMessage("EUR->KRW");
        }
    }
}

File: scala/src/test/scala/BankShould.scala
import domain.{Bank, Money}
import org.scalatest.EitherValues

class BankShould extends org.scalatest.funsuite.AnyFunSuite with EitherValues {
  private val bank: Bank = Bank().addExchangeRate("EUR", "USD", 1.2)

  test("10 EUR -> USD = 12 USD") {
    assert(bank.convert(Money(10, "EUR"), "USD").value === Money(12, "USD"))
  }

  test("10 EUR -> EUR = 10 EUR") {
    assert(bank.convert(Money(10, "EUR"), "EUR").value === Money(10, "EUR"))
  }

  test("Return a Left in case of missing exchange rates") {
    assert(bank.convert(Money(10, "EUR"), "KRW").left.value === "EUR->KRW")
  }

  test("Conversion with different exchange rates EUR -> USD") {
    assert(bank.convert(Money(10, "EUR"), "USD").value === Money(12, "USD"))
    assert(
      bank
        .addExchangeRate("EUR", "USD", 1.3)
        .convert(Money(10, "EUR"), "USD")
        .value === Money(13, "USD")
    )
  }
}
